dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/ParallelDependentTestFinder.java

```diff
@@ -0,0 +1,600 @@
+
+            // If only one half contains dependent tests, ignore all tests not in that half.
+            if (topResultsMatch) {
+                tests = topHalf;
+            } else {
+                tests = botHalf;
+            }
+
+            tests.removeAll(addOnTests);
+            tests.remove(tests.size() - 1);
+
+            topHalf = new LinkedList<>(tests.subList(0, tests.size() / 2));
+            botHalf = new LinkedList<>(tests.subList(tests.size() / 2, tests.size()));
+        }
+
+        if (!tests.isEmpty()) {
+            if (isOriginalOrder) {
+                addDependency(dependentTestName, tests.get(0));
+            } else {
+                addDependency(tests.get(0), dependentTestName);
+            }
+        }
+    }
+
+    /**
+     * @return A set of tests that must come AFTER the test in question.
+     */
+    private Set<String> getAfterDependencies(final String testName) {
+        return knownDependencies.entrySet().stream()
+                .filter(dependencies -> dependencies.getValue().contains(testName))
+                .map(Map.Entry::getKey)
+                .collect(Collectors.toSet());
+    }
+
+    public static List<String> dependenciesToString(final List<ParallelDependentTestFinder> dependencies) {
+        return dependencies.stream()
+                .map(ParallelDependentTestFinder::dependencyToString)
+                .reduce(new ArrayList<>(), (a, b) -> {
+                    a.addAll(b);
+                    return a;
+                });
+    }
+
+    public List<String> dependencyToString() {
+        final List<String> result = new ArrayList<>();
+
+        result.add(Constants.TEST_LINE + dependentTestName);
+        result.add("Intended behavior: " + originalOrder.results.get(dependentTestName));
+        // Says when executed before but uses getAfterDependencies because it must be executed
+        // before the tests that come after.
+        result.add("when executed before: " + getAfterDependencies(dependentTestName));
+        result.add("when executed after: " + knownDependencies.get(dependentTestName));
+        result.add("The revealed different behavior: " + newOrder.results.get(dependentTestName));
+
+        return result;
+    }
+
+    public static void writeToFile(final ParallelDependentTestFinder dtFinder) {
+        writeToFile(Collections.singletonList(dtFinder));
+    }
+
+    public static void writeToFile(final ParallelDependentTestFinder dtFinder, final File outputFile) {
+        writeToFile(Collections.singletonList(dtFinder), outputFile);
+    }
+
+    public static void writeToFile(final List<ParallelDependentTestFinder> dtFinders) {
+        writeToFile(dtFinders, DT_FILE);
+    }
+
+    /**
+     * Writes all the dependency information to a file.
+     * @param dtFinders The dtFinders that contain the dependency information we need.
+     */
+    public static void writeToFile(final List<ParallelDependentTestFinder> dtFinders, final File outputFile) {
```

Comment:
Instead of writing your own method to convert the precomputed dependencies to a string, please use the test data class here instead. https://github.com/winglam/dependent-tests-impact/pull/22/files/ba79a576feb562d4c089c3042b0e2f927a1f626e#diff-303b5f1678f338b5174bbfbdbfeef0b7R11

I will add a comment to Henry to delete his class after you have merged your PR.

You should also replace all uses of Strings to represent the precomputed dependencies with this test data class.

Reply:
I'm not sure how to incorporate this into my current program. Right now, I use a map to keep track of dependencies (each test maps to the tests that must be run before it). If I use this data class, both the test and it's dependencies will be stored in the class, so I won't have a separate key/value for the map. It isn't strictly necessary to use the map, but it seems like the most efficient option.

Additionally, that class doesn't seem to contain the correct information--are you suggesting I write my own version that has what I need (hence the "comment to Henry to delete his class...")?

Reply:
You can use the data class and a map by having the name of the dependent test be mapped to a set of the data class.

The class indeed does not contain the correct information. The only information that I think you would need is just what I added as a comment to Henry's PR.

BTW, you mention that you are only keeping track of which tests need to run before a dependent test. If that is true, double check your work to ensure that if a test suite containing T1 and T2 and the only dependency is that T2 must always come AFTER T1, then your work here will still terminate and output the correct result (one dependence saying T1 needs to come before T2). You may also want to imagine a slightly more difficult case and see whether your work will be able to handle such a case.

Reply:
I had been keeping track of which tests need to be run after by adding another entry to the map (saying that T1 has to come before T2, rather than T2 has to come after T1). I've just realized that this approach is probably not the best behavior, as it adds dependencies that aren't there.

In addition to the change you suggested on Henry's pull request, I think that I will have to add another field, either to specify whether the independent test needs to come before or after the dependent test, or include both a before test and after test in the data class. I think the best way to do this might be to make the data class hold a set of the before tests and a set of the after tests, then I could simply map each dependent test to a single instance of the data class.

Additionally, I think that we can omit the "when executed after: []" line that comes after the "revealed different behavior..." line. This is because the dependent test should be revealed any time one of the dependencies (either a test coming before or after) is not satisfied, right?

Reply:
"(saying that T1 has to come before T2, rather than T2 has to come after T1). I've just realized that this approach is probably not the best behavior, as it adds dependencies that aren't there."
I agree. I think the test data class should be improved to keep track of whether the dependence is a before or after dependence. 

"I think the best way to do this might be to make the data class hold a set of the before tests and a set of the after tests, then I could simply map each dependent test to a single instance of the data class."
Including a field for both and making them be sets sounds good to me.

"Additionally, I think that we can omit the "when executed after: []" line that comes after the "revealed different behavior..." line."
My suggestion would be that the test data class contains a set of before and a set of after tests but your map of precomputed dependences still maps from a dependent test to a set of test data classes. The reason for this implementation is that we can actually generate a "debug" log from this implementation. 

You are correct that the final product of the precomputed dependencies list would only need for each dependent test, which tests need to come before and after it. However, with just that information we lose the reason as to how we got those before and after tests and more importantly, what could have happened without them.

For example, the "debug" log might tell us something like T5 needs to come after T4 otherwise T4 will pass instead of fail compared to the original order. Another dependence of T4 might say T3 needs to come before T4 otherwise T4 will error instead of fail compared to the original order. These are two separate dependencies that are both necessary for T4 to have the same behavior as the original order but they affect T4 differently and with the debug log we can identify these cases. The debug log will also clearly explain to us why T4 needs T3 to come before it and T5 to come after as opposed to just knowing that T4 won't have the same behavior if those two tests are not in the right order.

To sum up, I would expect the internal representation to support the idea of the debug log and the tool to output the debug log for those that are curious about how the precomputed dependencies are generated. However, our tool should also generate a concise precomputed dependencies list where each dependent test only shows up once and the only fields needed are before and after tests. For the debug log, each dependent test may appear more than once and you would need before tests, after tests, the dependent test's behavior without those before and after tests, the test's behavior in the original order, and the test list that revealed that dependent tests behavior.

If you want more clarification, feel free to shoot me a text and come by the office.